parsers/CA_NS.py

```diff
@@ -1,213 +1,182 @@
-            "hydro": mix["Hydro"] / 100.0,
-            "wind": mix["Wind"] / 100.0,
-            "imports": mix["Imports"] / 100.0,
-        }
-
-        # datetime is in format '/Date(1493924400000)/'
-        # get the timestamp 1493924400 (cutting out last three zeros as well)
-        data_timestamp = int(mix["datetime"][6:-5])
-        data_date = datetime.fromtimestamp(data_timestamp, tz=timezone.utc)
-
-        # validate
-        valid = True
-        for gen_type, value in percent_mix.items():
-            percent_bounds = valid_percent[gen_type]
-            if not (percent_bounds[0] <= value <= percent_bounds[1]):
-                # skip this datapoint in the loop
-                valid = False
-                logger.warning(
-                    "discarding datapoint at {dt} due to {fuel} percentage "
-                    "out of bounds: {value}".format(
-                        dt=data_date, fuel=gen_type, value=value
-                    ),
-                    extra={"key": zone_key},
-                )
-        if not valid:
-            # continue the outer loop, not the inner
+def _parse_timestamp(timestamp: str) -> datetime:
+    """
+    Construct a datetime object from a date string formatted as, e.g.,
+    "/Date(1493924400000)/" by extracting the Unix timestamp 1493924400. Note
+    that the three trailing zeros are cut out as well).
+    """
+    return datetime.fromtimestamp(int(timestamp[6:-5]), tz=timezone.utc)
+
+
+def _get_ns_info(
+    session: Session, logger: Logger
+) -> tuple[ExchangeList, ProductionBreakdownList]:
+    # Request data from the source. Skip the first element of each JSON array
+    # because the reported base load is always 0 MW.
+    loads = {  # Lookup table mapping timestamps to base loads (in MW)
+        _parse_timestamp(load["datetime"]): load["Base Load"]
+        for load in session.get(LOAD_URL).json()[1:]
+    }
+    mixes = session.get(MIX_URL).json()[1:]  # Electricity mix breakdowns in %
+
+    exchanges = ExchangeList(logger)
+    production_breakdowns = ProductionBreakdownList(logger)
+    for mix in mixes:
+        timestamp = _parse_timestamp(mix["datetime"])
+
+        # Ensure the provided percentages are within bounds, similarly to the
+        # logic in https://www.nspower.ca/site/renewables/assets/js/site.js. In
+        # practical terms, I've seen hydro production go higher than 70%, which
+        # is way more than reported capacity.
+        if (
+            15 < mix["Biomass"]
+            or 60 < mix["Hydro"]
+            or 50 < mix["Imports"]
+            # The validation JS reports an error when Solid Fuel (coal) is over
+            # 85%, but as far as I can tell, that can actually be a valid
+            # result, I've seen it a few times. Use 98% instead.
+            or 98 < mix["Solid Fuel"]
+            or 55 < mix["Wind"]
+            # Gas
+            or 50 < mix["HFO/Natural Gas"] + mix["CT's"] + mix["LM 6000's"]
+        ):
```

Comment:
I would suggest removing these filters (I was the person who added them in the first place). They weren't accurate during last summer (datapoints were being ignored due to these filters), and that Javascript file no longer exists.

(actually this data doesn't seem to be shown anywhere on NS Power website anymore, but let's hope they keep the API endpoint up :hand_over_mouth: )

Reply:
Thanks for chiming in! Do "these filters" include the capacity filters (starting on line 103) as well, or only the percentage filters shown here?

Reply:
The percentage filters originally came in from that JS file, and the capacity filters were our own, I think. The capacity filters seem alright still, except I'm not sure if this endpoint reports Maritime Link import as part of the general hydro, in which case hydro could exceed the capacity. But we can leave it for now and check logs.